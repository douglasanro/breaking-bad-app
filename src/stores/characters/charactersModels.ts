import ICharacter from 'models/ICharacter';

export type TCharacterFilterStatus = 'Alive' | 'Dead' | '';

export type TFilterTypes = 'status';

export interface ICharacterFilter {
  status: TCharacterFilterStatus;
};

export interface ICharacterState {
  list: ICharacter[];
  hasNextPage: boolean;
  filter: ICharacterFilter;
  searchTerm: string;
};

export interface IFetchCharacters {
  limit: number;
  offset?: number;
};
