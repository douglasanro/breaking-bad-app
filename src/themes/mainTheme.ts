import { unstable_createMuiStrictModeTheme as createMuiTheme } from '@material-ui/core';

const mainTheme = createMuiTheme({
  palette: {
    primary: {
      light: '#69A704',
      main: '#3A7116',
      dark: '#064500',
      contrastText: '#FFF',
    },
    secondary: {
      light: '#FFFF51',
      main: '#DFD400',
      dark: '#A9A300',
      contrastText: '#000',
    },
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        "*, *::before, *::after": {
          boxSizing: "content-box",
        },
        body: {
          background: '#FFF'
        },
      },
    },
  },
});

export default mainTheme;
