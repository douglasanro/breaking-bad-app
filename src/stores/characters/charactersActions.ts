import { action, createAsyncAction } from 'typesafe-actions';
import { ICharacterFilter } from './charactersModels';
import ICharacter from 'models/ICharacter';

export enum ECharactersActions {
  FETCH_CHARACTERS_REQUEST = '@characters/FETCH_CHARACTERS_REQUEST',
  FETCH_CHARACTERS_SUCCESS = '@characters/FETCH_CHARACTERS_SUCCESS',
  FETCH_CHARACTERS_FAILURE = '@characters/FETCH_CHARACTERS_FAILURE',
  CLEAR_CHARACTERS = '@characters/CLEAR_CHARACTERS',
  HAS_NEXT_PAGE = '@characters/HAS_NEXT_PAGE',
  SET_SEARCH_TERM = '@characters/SET_SEARCH_TERM',
  SET_FILTER = '@characters/SET_FILTER'
}

export const setCharacters = createAsyncAction(
  ECharactersActions.FETCH_CHARACTERS_REQUEST,
  ECharactersActions.FETCH_CHARACTERS_SUCCESS,
  ECharactersActions.FETCH_CHARACTERS_FAILURE,
)<undefined, ICharacter[], Error>();

export const clearCharacters = () => action(ECharactersActions.CLEAR_CHARACTERS);

export const setHasNextPage = (flag: boolean) => action(ECharactersActions.HAS_NEXT_PAGE, flag);

export const setSearchTerm = (term: string) => action(ECharactersActions.SET_SEARCH_TERM, term);

export const setFilter = ({ status }: ICharacterFilter) => action(ECharactersActions.SET_FILTER, { status });
