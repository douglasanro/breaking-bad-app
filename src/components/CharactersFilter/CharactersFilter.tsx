import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import Box from '@material-ui/core/Box';
import { setFilter } from 'stores/characters/charactersActions';
import { rootState } from 'stores/rootStore';
import { ICharacterFilter } from 'stores/characters/charactersModels';
import ICharactersFilter from 'models/ICharactersFilter';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    box: {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(4),
    },
    formControl: {
      marginTop: theme.spacing(1),
      marginBottom: theme.spacing(1),
      minWidth: 150,
      '&:not(:last-child)': {
        marginRight: theme.spacing(1),
      },
    },
  })
);

const charactersFilters: ICharactersFilter[] = [
  {
    id: 'status',
    label: 'Status',
    initWithEmptyOption: true,
    items: [
      {
        id: 'alive',
        option: 'Alive',
        value: 'Alive',
      },
      {
        id: 'dead',
        option: 'Dead',
        value: 'Dead',
      },
    ],
  },
];

const CharactersFilter: React.FC = () => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const { filter } = useSelector(({ characters }: rootState) => characters);

  const handleChange = (event: React.ChangeEvent<{ name?: string | undefined; value: unknown }>) => {
    const filterId = event.target.name as keyof ICharacterFilter;
    dispatch(setFilter({ [filterId]: event.target.value as any }));
  };

  return (
    <Box className={classes.box}>
      <Typography>Filters:</Typography>
      {charactersFilters.map(({ id, label, initWithEmptyOption, items }) => (
        <FormControl key={id} variant="outlined" className={classes.formControl}>
          <InputLabel id={`characters-filter-${id}-label`}>{label}</InputLabel>
          <Select
            id={`characters-filter-${id}`}
            name={id}
            labelId={`characters-filter-${id}-label`}
            value={filter[id as keyof ICharacterFilter]}
            label={label}
            onChange={handleChange}
          >
            {initWithEmptyOption && (
              <MenuItem value="">
                <em>None</em>
              </MenuItem>
            )}
            {items.map(({ id, option, value }) => (
              <MenuItem key={id} value={value}>
                {option}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      ))}
    </Box>
  );
};

export default CharactersFilter;
