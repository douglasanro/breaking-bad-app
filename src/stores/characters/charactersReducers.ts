import { createReducer } from 'typesafe-actions';
import { ECharactersActions } from './charactersActions';
import { ICharacterState } from './charactersModels'

export const charactersInitialState: ICharacterState = {
  list: [],
  hasNextPage: true,
  filter: {
    status: ''
  },
  searchTerm: ''
};

const charactersReducers = createReducer(charactersInitialState, {
  [ECharactersActions.FETCH_CHARACTERS_SUCCESS]: (state, action) => ({
    ...state,
    list: [...new Set([...state.list, ...action.payload])]
  }),
  [ECharactersActions.CLEAR_CHARACTERS]: (state) => ({
    ...state,
    list: []
  }),
  [ECharactersActions.HAS_NEXT_PAGE]: (state, action) => ({
    ...state,
    hasNextPage: action.payload
  }),
  [ECharactersActions.SET_SEARCH_TERM]: (state, action) => ({
    ...state,
    searchTerm: action.payload
  }),
  [ECharactersActions.SET_FILTER]: (state, action) => ({
    ...state,
    filter: {...state.filter, ...action.payload}
  }),
});

export default charactersReducers;
