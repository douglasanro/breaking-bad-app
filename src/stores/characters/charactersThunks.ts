import { Dispatch } from 'redux';
import axios from 'axios';
import { setCharacters, setHasNextPage } from './charactersActions';
import { IFetchCharacters } from './charactersModels';
import ICharacter from 'models/ICharacter';
import { rootState } from 'stores/rootStore';

const baseUrl = 'https://breakingbadapi.com/api';
const endpoint = '/characters';
const defaultOffset = 0;

export const getCharacters = ({ limit, offset = defaultOffset }: IFetchCharacters) => async ( dispatch: Dispatch, getState: () => rootState) => {
  const { characters } = getState();
  const { searchTerm } = characters;
  try {
    dispatch(setCharacters.request());
    const searchByName = searchTerm ? `&name=${searchTerm}` : '';
    const { data } = await axios.get(`${baseUrl}${endpoint}?limit=${limit}&offset=${offset}${searchByName}`);

    if(!data.length) {
      dispatch(setHasNextPage(false));
      return;
    }

    const effectData = data.map(({ status, ...otherArgs}: ICharacter) => {
      const currentStatus = ['Alive', 'Dead'][Number(status !== 'Alive')];

      return {...otherArgs, status: currentStatus };
    })

    dispatch(setHasNextPage(true));
    dispatch(setCharacters.success(effectData));
  } catch (error) {
    dispatch(setCharacters.failure(error));
  }
};
