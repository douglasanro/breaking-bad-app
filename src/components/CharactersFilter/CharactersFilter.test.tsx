import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from '@material-ui/core/styles';
import CharactersFilter from './CharactersFilter';
import { charactersInitialState } from 'stores/characters/charactersReducers';
import mainTheme from 'themes/mainTheme';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
let store: ReturnType<typeof mockStore>;

const setupTest = () =>
  render(<CharactersFilter />, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </Provider>
    ),
  });

beforeEach(() => {
  store = mockStore({
    characters: charactersInitialState,
  });
});

test('Render CharactersFilter correctly', () => {
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});

test('Select status filter correctly', () => {
  const expectedActions = [
    { error: undefined, meta: undefined, payload: { status: 'Alive' }, type: '@characters/SET_FILTER' },
  ];
  const { getByTestId } = setupTest();
  fireEvent.change(getByTestId(/characters-filter-status/i), { target: { value: 'Alive' } });

  expect(store.getActions()).toEqual(expectedActions);
});
