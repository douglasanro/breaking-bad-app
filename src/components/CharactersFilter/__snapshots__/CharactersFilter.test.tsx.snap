// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CharactersFilter correctly 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-3 makeStyles-box-1"
  >
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Filters:
    </p>
    <div
      class="MuiFormControl-root makeStyles-formControl-2"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
        id="characters-filter-status-label"
      >
        Status
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="characters-filter-status-label characters-filter-status"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
          id="characters-filter-status"
          role="button"
          tabindex="0"
        >
          <span>
            â€‹
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          data-testid="characters-filter-status"
          name="status"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-6"
          >
            <span>
              Status
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
